File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Standings/Sources/AcoCarDriversSource.swift
import Foundation
import SwiftSoup

/// Pilotos por coche para WEC y Le Mans Cup — fuente COMPARTIDA porque ambas webs
/// (fiawec.com y lemanscup.com) usan exactamente la misma plantilla (mismo organizador,
/// ACO). Equivalente exacto de `AcoCarDriversSource.kt`. No conforma `StandingsSource`
/// (igual que en Android): esto alimenta `CarDriverModel`, no `StandingModel`.
///
/// HONESTO sobre el coste: hace falta una petición por coche además del listado — WEC
/// tiene ~35 coches, Le Mans Cup ~45. Se visitan con concurrencia acotada (8 a la vez).
/// Un fallo al visitar la ficha de un coche concreto solo deja a ESE coche sin pilotos,
/// nunca tumba la sincronización completa.
public final class AcoCarDriversSource: @unchecked Sendable {
    private let category: StandingsCategory
    /// Página con todos los coches de la temporada — fiawec.com/en/page/grid o
    /// lemanscup.com/en/car/{año}.
    private let listingUrl: String
    /// Texto del badge de clase (`span.fs-11`) a StandingsClass — se prueban en orden,
    /// las coincidencias más específicas ("LMP3 Pro/Am") deben ir antes que las
    /// genéricas ("LMP3").
    private let classMatchers: [(StandingsClass, (String) -> Bool)]

    public init(category: StandingsCategory, listingUrl: String, classMatchers: [(StandingsClass, (String) -> Bool)]) {
        self.category = category
        self.listingUrl = listingUrl
        self.classMatchers = classMatchers
    }

    // internal (no private): expuesta a test — ver parseCarRefs arriba.
    struct CarRef {
        var standingsClass: StandingsClass
        var carUrl: String
    }

    public func fetch(nowUtc: Date) async throws -> [CarDriverDraft] {
        let listingHtml = try await HTTPClient.fetchHTML(listingUrl)
        let carRefs = try parseCarRefs(listingHtml)

        let results = await withBoundedConcurrency(carRefs, limit: Self.concurrency) { ref in
            await self.fetchCarDrivers(carUrl: ref.carUrl, standingsClass: ref.standingsClass, nowUtc: nowUtc)
        }
        return results.flatMap { $0 }
    }

    // 04/09/2026 (Fase 1 del diagnóstico): separado de fetch() para poder testear la
    // resolución de clase por el badge (span.fs-11) y la extracción de la url de cada
    // coche contra un fixture HTML sin red — ver AcoCarDriversSourceTests. CarRef pasa a
    // internal por el mismo motivo.
    func parseCarRefs(_ html: String) throws -> [CarRef] {
        let listingDoc = try SwiftSoup.parse(html, listingUrl)
        let cards = try listingDoc.select("div.card-team").array()
        return try cards.compactMap { card -> CarRef? in
            let classText = try card.select("span.fs-11").first()?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard let standingsClass = classMatchers.first(where: { $0.1(classText) })?.0 else { return nil }
            guard let carUrl = try card.select("a.stretched-link").first()?.absUrl("href"), !carUrl.isEmpty else { return nil }
            return CarRef(standingsClass: standingsClass, carUrl: carUrl)
        }
    }

    private func fetchCarDrivers(carUrl: String, standingsClass: StandingsClass, nowUtc: Date) async -> [CarDriverDraft] {
        do {
            // El número de coche sale del propio tramo final de la URL de la ficha
            // (".../car/2026/35") — más fiable que un texto suelto.
            let trimmed = carUrl.hasSuffix("/") ? String(carUrl.dropLast()) : carUrl
            let carNumber = trimmed.components(separatedBy: "/").last ?? ""
            guard !carNumber.isEmpty, Int(carNumber) != nil else { return [] }

            let html = try await HTTPClient.fetchHTML(carUrl)
            return try parseCarPage(html, carUrl: carUrl, carNumber: carNumber, standingsClass: standingsClass, nowUtc: nowUtc)
        } catch {
            return []
        }
    }

    // 04/09/2026 (Fase 1 del diagnóstico): separado de fetchCarDrivers() para poder
    // testear el parsing de la ficha de coche (foto real vía <img src>, sin el truco de
    // carga perezosa que sí usa imsa.com) contra un fixture HTML sin red.
    func parseCarPage(_ html: String, carUrl: String, carNumber: String, standingsClass: StandingsClass, nowUtc: Date) throws -> [CarDriverDraft] {
        let doc = try SwiftSoup.parse(html, carUrl)
        let cards = try doc.select("a.card-driver").array()
        return try cards.compactMap { card -> CarDriverDraft? in
            let name = try card.select("div.py-4").first()?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !name.isEmpty else { return nil }
            let photoUrl = try card.select("img").first().flatMap { try? $0.absUrl("src") }.flatMap { $0.isEmpty ? nil : $0 }

            return CarDriverDraft(
                category: category,
                standingsClass: standingsClass,
                carNumber: carNumber,
                entryKey: TextNormalizer.normalize(name),
                name: name,
                photoUrl: photoUrl,
                updatedAtUtc: nowUtc
            )
        }
    }

    private static let concurrency = 8
}

File: PitBoard(Android)/app/src/main/java/com/pitboard/app/standings/sources/AcoCarDriversSource.kt
package com.pitboard.app.standings.sources

import com.pitboard.app.standings.CarDriverEntity
import com.pitboard.app.standings.StandingsCategory
import com.pitboard.app.standings.StandingsClass
import com.pitboard.app.standings.StandingsHttpClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import okhttp3.Request
import org.jsoup.Jsoup

/**
 * Pilotos por coche para WEC y Le Mans Cup — fuente COMPARTIDA porque ambas webs
 * (fiawec.com y lemanscup.com) usan exactamente la misma plantilla (mismo organizador,
 * ACO): comprobado a mano el 03/09/2026, byte a byte iguales en la parte que interesa
 * aquí — tarjetas `div.card-team` con la clase del coche (`span.fs-11`), su número (del
 * propio enlace `/en/car/{año}/{N}`, no hace falta leerlo de ningún texto) y el enlace a
 * su ficha; y en la ficha de cada coche, tarjetas `a.card-driver` con foto real (`img`
 * con `src` normal, sin el truco de carga perezosa que sí tiene imsa.com) y nombre
 * (`div.py-4`).
 *
 * HONESTO sobre el coste: como con ImsaStandingsSource, hace falta una petición por
 * coche además de la página de listado — WEC tiene ~35 coches, Le Mans Cup ~45. Se
 * visitan con concurrencia acotada (8 a la vez) para no golpear el servidor de golpe.
 * Un fallo al visitar la ficha de un coche concreto solo deja a ESE coche sin pilotos,
 * nunca tumba la sincronización completa.
 */
class AcoCarDriversSource(
    private val category: StandingsCategory,
    /** Página con todos los coches de la temporada — fiawec.com/en/page/grid o
     *  lemanscup.com/en/car/{año} (con el año ya sustituido). */
    private val listingUrl: String,
    /** Texto del badge de clase (`span.fs-11`, ej. "Hypercar", "LMP3 Pro/Am") a
     *  StandingsClass — se prueban en orden, así que las coincidencias más específicas
     *  ("LMP3 Pro/Am") deben ir antes que las genéricas ("LMP3"). */
    private val classMatchers: List<Pair<StandingsClass, (String) -> Boolean>>
) {

    suspend fun fetch(nowUtc: Long): List<CarDriverEntity> {
        val listingHtml = fetchHtml(listingUrl)
        val listingDoc = Jsoup.parse(listingHtml, listingUrl)

        data class CarRef(val standingsClass: StandingsClass, val carUrl: String)

        val carRefs = listingDoc.select("div.card-team").mapNotNull { card ->
            val classText = card.selectFirst("span.fs-11")?.text()?.trim().orEmpty()
            val standingsClass = classMatchers.firstOrNull { (_, matches) -> matches(classText) }?.first
                ?: return@mapNotNull null
            val carUrl = card.selectFirst("a.stretched-link")?.absUrl("href")?.takeIf { it.isNotBlank() }
                ?: return@mapNotNull null
            CarRef(standingsClass, carUrl)
        }

        return coroutineScope {
            carRefs.chunked(CONCURRENCY).flatMap { chunk ->
                chunk.map { ref -> async(Dispatchers.IO) { fetchCarDrivers(ref.carUrl, ref.standingsClass, nowUtc) } }
                    .awaitAll()
            }
        }.flatten()
    }

    private fun fetchCarDrivers(carUrl: String, standingsClass: StandingsClass, nowUtc: Long): List<CarDriverEntity> =
        runCatching {
            // El número de coche sale del propio tramo final de la URL de la ficha
            // (".../car/2026/35") — más fiable que leerlo de un texto suelto, que en
            // alguna de las dos webs va en un span aparte y en la otra pegado al
            // nombre del equipo.
            val carNumber = carUrl.trimEnd('/').substringAfterLast('/')
            if (carNumber.isBlank() || carNumber.toIntOrNull() == null) return@runCatching emptyList()

            val html = fetchHtml(carUrl)
            val doc = Jsoup.parse(html, carUrl)

            doc.select("a.card-driver").mapNotNull { card ->
                val name = card.selectFirst("div.py-4")?.text()?.trim().orEmpty()
                if (name.isBlank()) return@mapNotNull null
                val photoUrl = card.selectFirst("img")?.absUrl("src")?.takeIf { it.isNotBlank() }

                CarDriverEntity(
                    category = category,
                    standingsClass = standingsClass,
                    carNumber = carNumber,
                    entryKey = normalize(name),
                    name = name,
                    photoUrl = photoUrl,
                    updatedAtUtc = nowUtc
                )
            }
        }.getOrElse { emptyList() }

    private fun normalize(s: String): String =
        java.text.Normalizer.normalize(s, java.text.Normalizer.Form.NFD)
            .replace(Regex("\\p{Mn}+"), "")
            .lowercase()
            .replace(Regex("[^a-z0-9\\s]"), "")
            .replace(Regex("\\s+"), " ")
            .trim()

    private fun fetchHtml(url: String): String {
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "Mozilla/5.0 (Linux; Android 14) PitBoard/1.0")
            .build()
        return StandingsHttpClient.instance.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("$url: HTTP ${response.code}")
            response.body?.string() ?: error("$url: cuerpo vacío")
        }
    }

    companion object {
        private const val CONCURRENCY = 8
    }
}
